Class extension of proxy without construct trap fix #4729

rhuanjl opened the pull request:
Proxy trap function when there's no construct trap was always making a new object as the "this" to use - in the case of a super call this would break the class inheritance chain.

Currently CC's behaviour on this differers from jsc, v8 and sm. This fix brings CC into line with the others.

Should fix #4729

MSLaguana commented on test/es6/proxybugs.js:
```diff
@@ -383,6 +383,49 @@ var tests = [
+            var withtrap = new WithTrap();
+            assert.isTrue(withtrap.own);
+            assert.isTrue(withtrap.a());
+            assert.isUndefined(withtrap.trap);
```
Shouldn't this be `withtrap.noTrap`?

rhuanjl replied:
withtrap.trap was intentional - it's undefined as the "this" in the constructor trap is not the same this as the one for the target object.

I could add withtrap.noTrap as well, and I suppose I could make the trap do something to prove that the trap is working.

Key intent here was testing that the class extension is working correctly without picking up extra "this"s.

MSLaguana replied:
Oh, I misread what the test was doing. I thought you were verifying that using the proxy with the construct trap meant that it didn't actually call the original constructor function. I'm not sure I understand what you are trying to verify here; doesn't the `this` given to the proxy construct function get returned as the constructed object, so the `this.trap = true` should set the property on the instance?

rhuanjl replied:
I've expanded the case a bit further. I note that it covers more than it needs to - there were no tests of any kind (that I could see) for class that extend proxies so I thought I should do more than the specific point this patch fixes.

rhuanjl commented:
@dotnet-bot test OSX _no_jit_shared_osx_osx_test

boingoing commented on lib/Runtime/Library/JavascriptProxy.cpp:
```diff
@@ -2129,8 +2129,15 @@ namespace Js
+                    newThisObject = JavascriptOperators::NewScObjectNoCtor(targetObj, scriptContext);
+                    args.Values[0] = newThisObject;
+                }
+                else
```
I believe we don't need this else clause at all since `newThisObject` is not used again below. We only need to not overwrite `args.Values[0]`. The below loop is going to copy all the existing arguments into the `newValues` array anyway. In fact, it doesn't appear we even need to declare `newThisObject`. The above line could just say:

```c++
args.Values[0] = JavascriptOperators::NewScObjectNoCtor(targetObj, scriptContext);
```

sethbrenith replied:
It's used on line 2187: `aReturnValue = newThisObject;`

---
In reply to: [178616915](https://github.com/Microsoft/ChakraCore/pull/4908#discussion_r178616915) [](ancestors = 178616915)

boingoing replied:
Oh thanks for pointing this out. I searched for this name but it didn't highlight that one on line 2187. I see no problems with this code, then.

rhuanjl replied:
Could in theory replace aReturnValue below with args.Values[0]. If you want to remove a name from the code.

boingoing commented on test/es6/proxybugs.js:
```diff
@@ -383,6 +383,52 @@ var tests = [
+            
+            var withtrap = new WithTrap();
+            assert.isTrue(withtrap.own);
+            assert.throws(withtrap.a);
```
Why does this throw? Is it because `withtrap` doesn't connect up to the WithTrap prototype so it doesn't have the `a` method? In that case, should this be `assert.isUndefined` instead of `assert.throws` ?

sethbrenith replied:
Good point, this seems to be testing incorrect behavior. If this currently throws, then we should leave a commented `assert.isUndefined` and log another bug to follow up.

---
In reply to: [178618653](https://github.com/Microsoft/ChakraCore/pull/4908#discussion_r178618653) [](ancestors = 178618653)

rhuanjl replied:
That’s correct - when you have a trap it does not connect to the prototype (this felt weird to me but is the same behaviour as jsc, sm and v8)

Could indeed change to assert.isUndefined

rhuanjl commented:
@boingoing thanks for the review. I’m afraid I’m without a computer for the next week so I won’t be able to update this for a little while feel free to adopt it if you want to close it otherwise I will update next week.

rhuanjl commented:
@boingoing I've updated to latest and switched that test to assert.isUndefined instead of throws - it does feel odd to me BUT it's the same behaviour as v8, SM and jsc.

MSLaguana commented:
@dotnet-bot test OSX _no_jit_shared_osx_osx_test please
@dotnet-bot test Ubuntu shared_ubuntu_linux_test please
@dotnet-bot test Ubuntu static_ubuntu_linux_release please

rhuanjl commented:
Test failure seems to be a problem with CI rather than anything to do with this edit. Please let me know if/when you want me to update again or if you'd like me to do anything else.

MSLaguana commented:
@dotnet-bot test OSX _no_jit_shared_osx_osx test

MSLaguana commented:
@dotnet-bot test OSX _no_jit_shared_osx_osx test please

MSLaguana commented:
@dotnet-bot test OSX _no_jit_shared_osx_osx_test please

boingoing commented:
@rhuanjl The OSX test failures are a known failure in our CI and not due to your changes. If you could rebase this change on top of master to get it in-sync, I can marshal it through.

rhuanjl commented:
@boingoing rebased as requested

boingoing commented:
@rhuanjl Thanks for your contribution, this has landed in master.

jdalton commented:
Thanks! Do you all know which build of ChakraNode this will show up in?

MSLaguana commented:
You can get it from node-chakracore's `chakracore-master` branch if you want to build it yourself. The nightly builds of node-chakracore come from the `master` branch which tracks a stable chakracore branch, so it won't show up there until we make a new stable branch.
